src/vma/ib/mlx5/ib_mlx5_hw.cpp

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2001-2018 Mellanox Technologies, Ltd. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#if defined(DEFINED_DIRECT_VERBS) && (DEFINED_DIRECT_VERBS == 2)
+
+#include "vma/ib/mlx5/ib_mlx5.h"
+
+static int vma_ib_mlx5dv_get_qp(struct ibv_qp *qp, vma_ib_mlx5dv_qp_t *mlx5_qp);
+static int vma_ib_mlx5dv_get_cq(struct ibv_cq *cq, vma_ib_mlx5dv_cq_t *mlx5_cq);
+static int vma_ib_mlx5dv_get_rwq(struct ibv_exp_wq *wq, vma_ib_mlx5dv_rwq_t *mlx5_rwq);
+
+
+int vma_ib_mlx5dv_init_obj(vma_ib_mlx5dv_t *obj, uint64_t obj_type)
+{
+	int ret = 0;
+
+	if (obj_type & MLX5DV_OBJ_QP) {
+		ret = vma_ib_mlx5dv_get_qp(obj->dv.qp.in,
+				container_of(obj->dv.qp.out, vma_ib_mlx5dv_qp_t, dv));
+	}
+
+	if (!ret && (obj_type & MLX5DV_OBJ_CQ)) {
+		ret = vma_ib_mlx5dv_get_cq(obj->dv.cq.in,
+				container_of(obj->dv.cq.out, vma_ib_mlx5dv_cq_t, dv));
+	}
+
+	if (!ret && (obj_type & MLX5DV_OBJ_RWQ)) {
+		ret = vma_ib_mlx5dv_get_rwq(obj->dv.rwq.in,
+				container_of(obj->dv.rwq.out, vma_ib_mlx5dv_rwq_t, dv));
+	}
+
+	return ret;
+}
+
+static int vma_ib_mlx5dv_get_qp(struct ibv_qp *qp, vma_ib_mlx5dv_qp_t *mlx5_qp)
+{
+	int ret = 0;
+
+#if HAVE_DECL_IBV_MLX5_EXP_GET_QP_INFO
```

Comment:
Why we have if case here ?

Reply:
There are MOFED version w/o ibv_mlx5_exp_get_qp_info() function